Remove LinkedList usage in ClientMessage

ihsandemir opened the pull request:
Changed the ClientMessage so that it is no longer a LinkedList which may produce too much litter. 

Used next and previous pointer fields in Frame class to navigate between Frames. I had to write a new FrameIterator class which acts similar to ListIterator but with less litter. Regenerated all the codecs with this updated iterator.

One drawback was to shallow copy NULL_FRAME, BEGIN_FRAME, END_FRAME kind of static Frames which were not being copied before, since we need instance specific next and previous pointers.

Related protocol PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/236

sancar commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/ClientMessage.java:
```diff
@@ -137,47 +139,73 @@ private ClientMessage() {
+    }
+
+    public FrameIterator frameIterator() {
+        return new FrameIterator(startFrame);
```
Is it possible to remove FrameIterator also? ClientMessage could be the iterator itself. The drawback is you can not have two different iterators but I think we don't need it anyway.

ihsandemir replied:
OK, changed.

ihsandemir replied:
It came out that this was not a good idea. Some tests failed and when examined, I found out that the ClientMessage decode may happen from different executor threads and hence it is not thread safe if we keep the nextFrame variable in the ClientMessage which is being modified and read through different threads. Failing test: testGetStat_whenGetAsync  This line https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java#L91  the `DelegatingExecutionCallback` may decode the message using different executors.

Hence, I am reverting back the changes.

Scooletz commented:
> One drawback was to shallow copy NULL_FRAME, BEGIN_FRAME, END_FRAME kind of static Frames which were not being copied before, since we need instance specific next and previous pointers.

Yes, that the drawback when we make the `Frame` responsible for navigation. I was considering this for csharp client but I hit the same obstacle. Still, this PRs removes all the litter from the `LinkedList` nodes 😄  For csharp client I eventually decided to use a mix of fields and an array as shown in here:

https://github.com/hazelcast/hazelcast-csharp-client/blob/5627961eed4e092aa7611a9aeef714ad6536bf42/Hazelcast.Net/Hazelcast.Client.Protocol/ClientMessage.cs#L60-L62

As `Frame` is now `ValueType`, the only additional allocation is an array of frames. Still, it allows indexed based navigation. Additionally (and this requires benchmarking), a few more fields could be used to lower the probability of allocating the array altogether by introducing multiple `_head` fields. The getter would not change that much: 

https://github.com/hazelcast/hazelcast-csharp-client/blob/5627961eed4e092aa7611a9aeef714ad6536bf42/Hazelcast.Net/Hazelcast.Client.Protocol/ClientMessage.cs#L118-L123

For `Java`, as every frame is a heap allocated object the array approach would not change that much, the array still needs to be allocated and resized if needed, but again we could try to calculate if there are some possible gains. Same with adding a few `Frame` fields to the client message to make it a bit fatter. This would remove `next/prev` navigation from the frame (2x reference) and add just one in the `ClientMessage` (a specific `_frameN` field) + the tail array.

I haven't done all the calculations, but if this looks somehow interesting to you, I can do it.

ihsandemir commented:
> As `Frame` is now `ValueType`, the only additional allocation is an array of frames. Still, it allows indexed based navigation. Additionally (and this requires benchmarking), a few more fields could be used to lower the probability of allocating the array altogether by introducing multiple `_head` fields. The getter would not change that much:

I think that introducing an array which will keep pointers to each frame may not benefit much. We really do not need index based access if you look at the implementation and I removed it. Arrays would be good only if we had fixed frame sizes and I could have an array of frames (consecutive memory locations) and that array could be stack allocated in other languages. But our Frame sizes are varying a lot and hence it does not look feasible.

We can always test your suggestion for performance and littering but I doubt that it would do any good at least for Java side.

ihsandemir commented:
verify

ihsandemir commented:
run-lab-run
